src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -170,6 +171,99 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
+        /// <param name="expected">The expected lock file structure. Usuaully generated from the project.</param>
+        /// <param name="actual">The lock file that was loaded from the file on disk.</param>
+        /// <returns>A ValueTuple containing two fields. First, a boolean to represent whether the lock file is still
+        /// valid (structure of the two lock files are the same). Second, a list of matched dependencies if the lock
+        /// file is still valid, otherwise null.</returns>
+        public static (bool lockFileStillValid, List<KeyValuePair<LockFileDependency, LockFileDependency>> matchedDependencies)
+            IsLockFileStillValid(PackagesLockFile expected, PackagesLockFile actual)
+        {
+            // do quick checks for obvious structure differences
+            if (expected.Version != actual.Version)
+            {
+                return (false, null);
+            }
+
+            if (expected.Targets.Count != actual.Targets.Count)
+            {
+                return (false, null);
+            }
+
+            foreach (var expectedTarget in expected.Targets)
+            {
+                PackagesLockFileTarget actualTarget = null;
+
+                for (var i = 0; i < actual.Targets.Count; i++)
+                {
+                    if (actual.Targets[i].TargetFramework == expectedTarget.TargetFramework)
+                    {
+                        if (actualTarget == null)
+                        {
+                            actualTarget = actual.Targets[i];
+                        }
+                        else
+                        {
+                            // more than 1? possible bug or bad hand edited lock file.
+                            return (false, null);
+                        }
+                    }
+
+                    if (actualTarget == null)
+                    {
+                        return (false, null);
+                    }
+
+                    if (actualTarget.Dependencies.Count != expectedTarget.Dependencies.Count)
+                    {
+                        return (false, null);
+                    }
+                }
+            }
+
+            // no obvious structure difference, so start trying to match individual dependencies
+            var matchedDependencies = new List<KeyValuePair<LockFileDependency, LockFileDependency>>();
+            var lockFileStillValid = true;
+            var dependencyComparer = LockFileDependencyComparerWithoutContentHash.Default;
+
+            foreach (var expectedTarget in expected.Targets)
+            {
+                var actualTarget = actual.Targets.Single(t => t.TargetFramework == expectedTarget.TargetFramework);
+
+                // Duplicate dependencies list so we can remove matches to validate that all dependencies were matched
+                var actualDependencies = new Dictionary<LockFileDependency, LockFileDependency>(
```

Comment:
I don't understand why you use a dictionary when a simple `HashSet<LockFileDependency>` would suffice.

Reply:
I tried to at first, but netstandard2.0 doesn't have `TryGetValue`, or any other reasonable way to get the value. Checking `Contains` is insufficient, because the equality comparer intentionally ignores content hash, so after checking the key exists, we really need the value so we can compare the two dependencies content hashes.

Reply:
Feels to me that the LockFileDependency class is overused. 
The "this is my input", and "this is my lock file" models should be different.